specs/capella/beacon-chain.md

```diff
@@ -323,11 +330,17 @@ def process_withdrawals(state: BeaconState, payload: ExecutionPayload) -> None:
     for expected_withdrawal, withdrawal in zip(expected_withdrawals, payload.withdrawals):
         assert withdrawal == expected_withdrawal
         decrease_balance(state, withdrawal.validator_index, withdrawal.amount)
-    if len(expected_withdrawals) > 0:
-        latest_withdrawal = expected_withdrawals[-1]
-        state.next_withdrawal_index = WithdrawalIndex(latest_withdrawal.index + 1)
-        next_validator_index = ValidatorIndex((latest_withdrawal.validator_index + 1) % len(state.validators))
-        state.next_withdrawal_validator_index = next_validator_index
+
+    if len(expected_withdrawals) == 0:
+        # advance sweep if there were no withdrawals
+        next_validator_index = (state.next_withdrawal_validator_index + MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP) % len(state.validators)
+        state.next_withdrawal_validator_index = ValidatorIndex(next_validator_index)
+        return
+
+    latest_withdrawal = expected_withdrawals[-1]
+    state.next_withdrawal_index = WithdrawalIndex(latest_withdrawal.index + 1)
+    next_validator_index = ValidatorIndex((latest_withdrawal.validator_index + 1) % len(state.validators))
+    state.next_withdrawal_validator_index = next_validator_index
```

Comment:
Should be changed to always move the pointer forward. For the case of 0 withdrawals, incomplete withdrawals, and max withdrawals.

```py
    if len(expected_withdrawals) < MAX_WITHDRAWALS_PER_PAYLOAD:
        next_validator_index = (state.next_withdrawal_validator_index + MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP) % len(state.validators)
    else:
        next_validator_index = ValidatorIndex((expected_withdrawals[-1].validator_index + 1) % len(state.validators))

    state.next_withdrawal_validator_index = next_validator_index
    state.next_withdrawal_index = WithdrawalIndex(state.next_withdrawal_index + len(expected_withdrawals))
```

Reply:
I don't understand why the full sweep can only happen in the 0 withdrawal case and not in the incomplete withdrawals case. You be scanning a range of validators twice in consecutive blocks due to not updating the pointer position.